systemd_ctypes/bus.py

```diff
@@ -220,3 +223,125 @@ def add_match(self, rule, handler):
         slot = Slot(handler)
         super().add_match(byref(slot), rule, slot.callback, slot.userdata)
         return slot
+
+    def add_object(self, path, obj):
+        obj.prepare()
+        slot = Slot(obj.message_received)
+        super().add_object(byref(slot), path, slot.callback, slot.userdata)
+        return slot
+
+
+class BaseObject:
+    @classmethod
+    def prepare(cls):
+        raise NotImplementedError
+
+    def message_received(self, message):
+        try:
+            reply = message.new_method_return()
+            handled = self.handle_method_call(message, reply)
+        except BusError:
+            assert False  # needs proper exception handling
+
+        if handled:
+            # message_send() is not available in RHEL 8
+            message.get_bus().send(reply, None)
+            return 1
+        else:
+            return 0
+
+    def handle_method_call(self, message, reply):
+        raise NotImplementedError
+
+
+class Object(BaseObject):
+    _interface_name = None
+    _property_info = None
+    _method_info = None
+
+    @staticmethod
+    def interface(name):
+        def wrapper(cls):
+            cls._interface_name = name
+            return cls
+        return wrapper
+
+    @staticmethod
+    def property(type, name=None):
+        def wrapper(f):
+            f._export_type = 'property'
+            f._property_name = name
+            f._property_type = type
+            return f
+        return wrapper
+
+    @staticmethod
+    def method(out_signature, name, in_signature=''):
+        def wrapper(f):
+            f._export_type = 'method'
+            f._method_name = name
+            f._method_in_signature = in_signature
+            f._method_out_signature = out_signature
+            return f
+        return wrapper
+
+    @classmethod
+    def prepare(cls):
+        if cls._property_info is not None:
+            return
+
+        cls._property_info = {}
+        cls._method_info = {}
+
+        for name, value in cls.__dict__.items():
+            if name.startswith('__') or not hasattr(value, '_export_type'):
+                continue
+
+            if value._export_type == 'property':
+                cls._property_info[value._property_name] = value
+            elif value._export_type == 'method':
+                cls._method_info[value._method_name] = value
+
+    def append_property(self, reply, handler):
+        reply.open_container(ord('v'), handler._property_type)
+        reply.append_arg(handler._property_type, handler(self))
+        reply.close_container()
+
+    def handle_method_call(self, message, reply):
+        interface = message.get_interface(message)
+        method = message.get_member()
+
+        if interface == 'org.freedesktop.DBus.Introspectable':
```

Comment:
Ah, I've missed The Big Switch™ since the days of TurboVision :grin:

Reply:
naja...

You actually find the one spot where this approach suffers the most: we have to return a single XML document for introspection.  Multiple interfaces at the same object path could otherwise peacefully co-exist (and even properties would work correctly) but the XML is one document, generated in one spot.

Also TODO: I need to actually generate that document :rofl:

Reply:
Don't worry, that was just a reminiscence of the past, not any suggestion to change it.